File: python/gsvolley.py
#!/usr/bin/python

import sys

import pygtk
pygtk.require("2.0")
import gtk
import glib
import cairo
import math
import random

WIN_WIDTH  = 640
WIN_HEIGHT = 400

GRAVITY           = 0.03
SCORE_INIT        = 2

BALL_RADIUS       = 10
BALL_SERVE_OFFSET = 76.5
BALL_Y_INIT = 20
PENALTY_RADIUS    = 30

SLIME_RADIUS = 40

class Ball():
    def __init__(self):
        self.init(Slime.BLUE)

    def init(self, team):
        if team == Slime.BLUE:
            self.x = BALL_SERVE_OFFSET
        else:
            self.x = WIN_WIDTH - BALL_SERVE_OFFSET

        self.y = BALL_Y_INIT
        self.vx = 0
        self.vy = 0


    def move(self):
        self.vy += GRAVITY
        self.x += self.vx
        self.y += self.vy

        if self.x < 0 + BALL_RADIUS:
            self.x = (BALL_RADIUS - self.x) + BALL_RADIUS
            self.vx = - self.vx
        if self.x > WIN_WIDTH - BALL_RADIUS:
            self.x = (WIN_WIDTH - BALL_RADIUS) - (self.x - (WIN_WIDTH - BALL_RADIUS))
            self.vx = - self.vx

class Vector():
    def __init__(self, x = 0, y = 0):
        self.x = x
        self.y = y

    @staticmethod
    def inner_product(a,b):
        return a.x * b.x + a.y * b.y

    @staticmethod
    def outer_product(a,b):
        return a.x * b.y - b.x * a.y

    def norm(self):
        return math.sqrt(self.x*self.x + self.y*self.y)



class GameInfo():
    (GS_INIT, GS_PLAY,  GS_GOT_BY_BLUE,
     GS_GOT_BY_RED,  GS_SERVICE_BY_BLUE,
     GS_SERVICE_BY_RED, GS_WON_BY_BLUE,
     GS_WON_BY_RED) = range(8)

    def __init__(self):
        self.slime_blue = None
        self.slime_red = None
        self.ball = None
        self.window = None


    def init(self):
        self.ball.init(Slime.BLUE)
        self.slime_blue.init()
        self.slime_red.init()
        
        self.ball_owner = Slime.BLUE
        self.ball_count = 1
        self.wait_count = 0
        self.score_blue = SCORE_INIT
        self.score_red = SCORE_INIT
        self.state = self.GS_SERVICE_BY_BLUE

    def score_move(self, win):
        if self.ball_count > 0:
            if self.wait_count > 0:
                self.wait_count -= 1
            else:
                self.ball_count -= 1
                if self.ball_count == 0:
                    self.wait_count = 50
                else:
                    self.wait_count = 7
                if win == Slime.BLUE:
                    self.score_red -= 1
                    if self.score_red <= 0:
                        self.state = GameInfo.GS_WON_BY_BLUE
                else:
                    self.score_blue -= 1
                    if self.score_blue <= 0:
                        self.state = GameInfo.GS_WON_BY_RED
        else:
            if self.wait_count > 0 :
                self.wait_count -= 1
            else:
                if win == Slime.BLUE:
                    self.state = GameInfo.GS_SERVICE_BY_BLUE
                else:
                    self.state = GameInfo.GS_SERVICE_BY_RED

    def serve_set(self, win):
        if win == Slime.BLUE:
            self.ball.init(Slime.BLUE)
        else:
            self.ball.init(Slime.RED)
        self.slime_blue.init()
        self.slime_red.init()
        self.penalty_y = random.uniform(120, WIN_HEIGHT -  40)
        self.ball_count = 1
        self.wait_count = 0
        self.state = GameInfo.GS_PLAY        

    def collision_ball_slime(self, team):
        b = self.ball
        if team == Slime.BLUE:
            s = self.slime_blue
        else:
            s = self.slime_red
        
        v_s = Vector(b.vx, b.vy)
        v_a = Vector(s.x - b.x, WIN_HEIGHT - b.y)
        v_b = Vector(s.x - (b.x + b.vx), WIN_HEIGHT - (b.y + b.vy))
        c = Vector.inner_product(v_a, v_s) * Vector.inner_product(v_b, v_s)
        try:
            d = abs(Vector.outer_product(v_s, v_a))/ v_s.norm()
        except ZeroDivisionError:
            return
        r = SLIME_RADIUS + BALL_RADIUS
        if d <= r and ( c<=0 or r>v_a.norm() or r > v_b.norm()):
            dx_0 = b.x - s.x
            dy_0 = b.y - WIN_HEIGHT
            dist_0 = Vector(dx_0, dy_0).norm()
            angle = math.acos(dx_0/dist_0)
            cos2x = math.cos(2*angle)
            sin2x = math.sin(2*angle)
            b.vx, b.vy = (-b.vx * cos2x + b.vy * sin2x,
                          b.vx * sin2x + b.vy * cos2x)

            self.ball_owner = team
            self.ball_count += 1

    def collision_penalty(self):
        b = self.ball
        if (Vector(b.x - WIN_WIDTH/2, b.y - self.penalty_y).norm() <
            PENALTY_RADIUS + BALL_RADIUS):
            self.wait_count -= 1
            if self.ball_owner == Slime.BLUE:
                if self.score_blue > 1 and self.wait_count <= 0:
                    self.score_blue -= 1
                    self.wait_count = 4
            else:
                if self.score_red > 1 and self.wait_count <= 0:
                    self.score_red -= 1
                    self.wait_count = 4
                    
            

class GameWin():
    def __init__(self, window):
        self.window = window
        self.cr = window.cairo_create()
        self.cr.set_font_size(16)

        self.pattern_blue = cairo.SolidPattern(0,0,1)
        self.pattern_red = cairo.SolidPattern(1,0,0)
        self.pattern_ball = cairo.SolidPattern(1,1,1)
        self.pattern_penalty = cairo.SolidPattern(0,1,0)

    def draw_back(self):
        self.cr.set_source_rgb(0,0,0)
        self.cr.rectangle(0,0,WIN_WIDTH, WIN_HEIGHT)
        self.cr.fill()

        self.cr.set_source_rgb(1,1,1)
        self.cr.rectangle(WIN_WIDTH/2 - 1, 120, 2, WIN_HEIGHT)
        self.cr.fill()

    def draw_slime(self, slime):
        if slime.team == Slime.BLUE:
            self.cr.set_source(self.pattern_blue)
        else:
            self.cr.set_source(self.pattern_red)
        self.cr.move_to(slime.x, WIN_HEIGHT)
        self.cr.arc(slime.x, WIN_HEIGHT, SLIME_RADIUS,
               math.pi, 0)
        self.cr.fill()

    def draw_ball(self, ball):
        self.cr.set_source(self.pattern_ball)
        self.cr.arc(ball.x, ball.y, BALL_RADIUS,
                    0, 2 * math.pi)
        self.cr.fill()

    def draw_penalty(self, y):
        self.cr.set_line_width(5)
        self.cr.set_source(self.pattern_penalty)
        self.cr.arc(WIN_WIDTH/2, y, PENALTY_RADIUS, 0, 2 * math.pi)
        self.cr.stroke_preserve()
        
        self.cr.set_source_rgb(0,0,0)
        self.cr.fill()

        self.cr.set_source(self.pattern_penalty)
        self.cr.arc(WIN_WIDTH/2, y, PENALTY_RADIUS - 15, 0, 2 * math.pi)
        self.cr.set_line_width(5)
        self.cr.stroke_preserve()

        self.cr.set_source_rgb(0,0,0)
        self.cr.fill()

    def draw_score(self, blue, count, red):
        buf = "Blue: %2d Ball: %3d Red: %2d"% (blue, count,red)
        self.show_text_centering(buf, 20)

    def show_text_centering(self, text, y):
        te = self.cr.text_extents(text)
        self.cr.move_to((WIN_WIDTH-te[2])/2, y)
        self.cr.show_text(text)

class Slime():
    BLUE, RED = range(2)
    def __init__(self, team):
        self.team = team
        if self.team == Slime.BLUE:
            self.limit_left = SLIME_RADIUS
            self.limit_right = WIN_WIDTH / 2 - SLIME_RADIUS
        else:
            self.limit_left = WIN_WIDTH / 2 + SLIME_RADIUS
            self.limit_right = WIN_WIDTH  - SLIME_RADIUS
        self.init()
            
    def init(self):
        if self.team == Slime.BLUE:
            self.x = self.limit_left
            self.vx = 0
        else:
            self.x = self.limit_right
            self.vx = 0

    def move(self):
        self.x += self.vx
        if self.x < self.limit_left:
            self.x = self.limit_left
            self.vx = 0
        if self.x > self.limit_right:
            self.x = self.limit_right
            self.vx = 0

def cb_expose_event(widget, event, gameinfo):
    gw = GameWin(widget.get_window())
    gw.draw_back()
    gw.draw_slime(gameinfo.slime_blue)
    gw.draw_slime(gameinfo.slime_red)
    gw.draw_penalty(gameinfo.penalty_y)
    gw.draw_ball(gameinfo.ball)
    gw.draw_score(gameinfo.score_blue, gameinfo.ball_count,
                  gameinfo.score_red)
    
    if gameinfo.state == GameInfo.GS_INIT:
        pass # OK
    elif gameinfo.state == GameInfo.GS_PLAY:
        pass # OK
    elif gameinfo.state == GameInfo.GS_GOT_BY_BLUE:
        gw.show_text_centering("DROPPED BY RED", 40)
    elif gameinfo.state == GameInfo.GS_GOT_BY_RED:
        gw.show_text_centering ("DROPPED BY BLUE", 40);
    elif gameinfo.state == GameInfo.GS_SERVICE_BY_BLUE:
        pass # OK
    elif gameinfo.state == GameInfo.GS_SERVICE_BY_RED:
        pass # OK
    elif gameinfo.state == GameInfo.GS_WON_BY_BLUE:
        gw.show_text_centering ("PRESS R TO PLAY AGAIN", 80);          
        gw.show_text_centering ("WON BY BLUE", 60);    
    elif gameinfo.state == GameInfo.GS_WON_BY_RED:
        gw.show_text_centering ("PRESS R TO PLAY AGAIN", 80);          
        gw.show_text_centering ("WON BY RED", 60);    
    else:
        raise RuntimeError("ERROR!! Unknown Game State")
    return False


def cb_key_press_event(widget, event, gameinfo):
    key = gtk.gdk.keyval_name(event.keyval)
    if  key == 'a':
        gameinfo.slime_blue.vx = -1
    elif key == 's':
        gameinfo.slime_blue.vx = 1
    elif key == 'Left':
        gameinfo.slime_red.vx = -1
    elif key == 'Right':
        gameinfo.slime_red.vx = 1
    elif key == 'r':
        if (gameinfo.state == GameInfo.GS_WON_BY_BLUE or
            gameinfo.state == GameInfo.GS_WON_BY_RED):
            gameinfo.state = GameInfo.GS_INIT
    else:
        pass # OK
    return False

def event_loop(gameinfo):
    if gameinfo.window.get_window() is None:
        return False
    
    if gameinfo.state == GameInfo.GS_INIT:
        gameinfo.init()
    elif gameinfo.state == GameInfo.GS_PLAY:
        gameinfo.slime_blue.move()
        gameinfo.slime_red.move()
        gameinfo.collision_ball_slime(Slime.BLUE)
        gameinfo.collision_ball_slime(Slime.RED)
        gameinfo.collision_penalty()
        gameinfo.ball.move()

        if gameinfo.ball.y >= WIN_HEIGHT - BALL_RADIUS:
            if gameinfo.ball.x <= WIN_WIDTH / 2:
                gameinfo.state = GameInfo.GS_GOT_BY_RED
            else:
                gameinfo.state = GameInfo.GS_GOT_BY_BLUE

    elif gameinfo.state == GameInfo.GS_GOT_BY_BLUE:
        gameinfo.score_move(Slime.BLUE)
    elif gameinfo.state == GameInfo.GS_GOT_BY_RED:
        gameinfo.score_move(Slime.RED)
    elif gameinfo.state == GameInfo.GS_SERVICE_BY_BLUE:
        gameinfo.serve_set(Slime.BLUE)
    elif gameinfo.state == GameInfo.GS_SERVICE_BY_RED:
        gameinfo.serve_set(Slime.RED)
    elif gameinfo.state == GameInfo.GS_WON_BY_BLUE:
        pass # OK
    elif gameinfo.state == GameInfo.GS_WON_BY_RED:
        pass # OK
    else:
        raise RuntimeError("ERROR!! Unknown Game State\n")

    gameinfo.window.queue_draw()
    return True

def cb_delete_event(widget, event, timeout):
    if timeout:
        glib.source_remove(timeout)
    return False

def main():
    window = gtk.Window()

    gameinfo = GameInfo()
    gameinfo.slime_blue = Slime(Slime.BLUE)
    gameinfo.slime_red =  Slime(Slime.RED)
    gameinfo.ball = Ball()
    gameinfo.window = window
    gameinfo.init()

    window.set_size_request(WIN_WIDTH, WIN_HEIGHT)
    
    window.set_resizable(False)
    window.set_app_paintable(True)
    timeout = glib.timeout_add(10, event_loop, gameinfo)
    window.connect("expose_event", cb_expose_event, gameinfo)
    window.connect("key_press_event", cb_key_press_event, gameinfo)
    window.connect("delete_event", cb_delete_event, timeout)
    window.connect("destroy", gtk.main_quit)
    window.show_all()
    gtk.main()

if __name__ == "__main__":
    main()

File: haskell/gsvolley.hs
module Main where 
import Control.Monad.Trans(liftIO)
import Graphics.UI.Gtk
import Graphics.Rendering.Cairo
import Data.IORef


-- ball
ball_y_init = 20 ::Double

win_width  = 640 :: Double
win_height = 400 :: Double
slime_radius = 40 :: Double
gravity    = 0.03

score_init = 50 :: Int

ball_radius =    10 :: Double
ball_serve_offset = 76.5
penalty_radius = 30 :: Double


data Ball = Ball {
  ball_x :: Double,
  ball_y :: Double,
  ball_vx :: Double,
  ball_vy :: Double
  } deriving(Show)
       
ball_new t =
  Ball {
    ball_x =  if t == Blue then
                ball_serve_offset
              else
                win_width - ball_serve_offset,
    ball_y =  ball_y_init,
    ball_vx = 0,
    ball_vy = 0
    }
      
ball_move b@Ball{ball_x=x, 
                 ball_y=y,
                 ball_vx=vx,
                 ball_vy=vy} = 
  reflection b{ball_x = x + vx,
               ball_y = y + vy,
               ball_vy = vy + gravity}
  where reflection b@Ball{ball_x = x, ball_vx = vx} 
          | x < 0 + ball_radius = b{ ball_x = (ball_radius - x) + ball_radius,
                                   ball_vx = -vx}
          | x > win_width - ball_radius =
            b{ball_x = (win_width - ball_radius) - 
                       (x - (win_width - ball_radius)),
              ball_vx = - vx}
          | otherwise = b
                                
data GameState = GS_INIT 
               | GS_PLAY 
               | GS_GOT_BY_BLUE 
               | GS_GOT_BY_RED  
               | GS_SERVICE_BY_BLUE
               | GS_SERVICE_BY_RED
               | GS_WON_BY_BLUE
               | GS_WON_BY_RED 
                 deriving(Show, Eq)

data GameInfo = GameInfo {
  gi_slime_blue :: Slime,
  gi_slime_red :: Slime,
  gi_ball :: Ball,
  --  window :: GtkWidget
  gi_state :: GameState,
  gi_score_blue :: Int,
  gi_score_red :: Int,
  gi_ball_count :: Int,
  gi_ball_owner :: Team,
  gi_penalty_y :: Double,
  gi_wait_count :: Int
  } deriving(Show)
            
gameinfo_new = GameInfo {
  gi_slime_blue = slime_new Blue,
  gi_slime_red  = slime_new Red,
  gi_ball = ball_new Blue,
  gi_state = GS_INIT,
  gi_score_blue = score_init,
  gi_score_red  = score_init,
  gi_ball_count = 0,
  gi_ball_owner = Blue,
  gi_penalty_y  = 0,
  gi_wait_count = 0
  }

data GameWin = GameWin {
  
  } deriving(Show)
           
data Team = Blue
          | Red
          deriving(Show, Eq)

data Slime = Slime {
  slime_team :: Team,
  slime_x :: Double,
  slime_vx :: Double,
  slime_limit_left ::  Double,
  slime_limit_right :: Double
  } deriving(Show)

slime_new team = slime_init Slime {
  slime_team = team,
  slime_x  = 0,
  slime_vx = 0,
  slime_limit_left = if team == Blue then
                       slime_radius
                     else
                       win_width / 2 + slime_radius,
  slime_limit_right = if team == Blue then
                        win_width / 2 - slime_radius
                      else
                        win_width - slime_radius
  }

slime_init Slime{
  slime_team=team,
  slime_x=x,
  slime_vx=vx,
  slime_limit_left=limit_left,
  slime_limit_right=limit_right
  } = Slime {
  slime_team=team,
  slime_x = if team == Blue then 
              slime_radius
            else 
              win_width - slime_radius,
  slime_vx=0,
  slime_limit_left=limit_left,
  slime_limit_right=limit_right
  }

slime_move s@Slime{slime_x=x,
                   slime_vx=vx,
                   slime_limit_left=limit_left,
                   slime_limit_right=limit_right} = 
  let x2 = x + vx in
  s {slime_x = case () of 
          _ | x2 < limit_left  -> limit_left
            | x2 > limit_right -> limit_right
            | otherwise  -> x2,
     slime_vx = if x2 < limit_left || x2 > limit_right then
                  0
                else
                  vx}
                   

pattern_blue = setSourceRGB 0 0 1
pattern_red = setSourceRGB 1 0 0 
pattern_ball = setSourceRGB 1 1 1
pattern_penalty = setSourceRGB 0 1 0

cb_expose_event iogameinfo = do
  drawWin <- eventWindow
  liftIO $ do
    gameinfo <- readIORef iogameinfo
    renderWithDrawable drawWin $ do
      setFontSize 16
      draw_back
      draw_slime $ gi_slime_blue gameinfo
      draw_slime $ gi_slime_red gameinfo
      draw_penalty $ gi_penalty_y gameinfo
      draw_ball $ gi_ball gameinfo
      draw_score (gi_score_blue gameinfo) (gi_ball_count gameinfo) 
        (gi_score_red gameinfo) 
      case (gi_state gameinfo) of
        GS_INIT -> return ()
        GS_PLAY -> return ()
        GS_GOT_BY_BLUE -> show_text_centering "DROPPED BY RED" 40
        GS_GOT_BY_RED -> show_text_centering "DROPPED BY BLUE" 40
        GS_SERVICE_BY_BLUE -> return ()
        GS_SERVICE_BY_RED  -> return ()
        GS_WON_BY_BLUE -> do
          show_text_centering "PRESS R TO PLAY AGAIN" 80
          show_text_centering "WON BY BLUE" 60
        GS_WON_BY_RED -> do
          show_text_centering "PRESS R TO PLAY AGAIN" 80
          show_text_centering "WON BY RED" 60
      return True
  where
    draw_back = do
      setSourceRGB 0 0 0
      rectangle 0 0 win_width win_height
      fill
      setSourceRGB 1 1 1
      rectangle (win_width/2 - 1) 120 2 win_height
      fill
    draw_slime s = do
      if (slime_team s) == Blue then
        do 
          pattern_blue
        else
        do
          pattern_red
      moveTo (slime_x s) win_height
      arc (slime_x s) win_height slime_radius (1 * pi) 0
      fill
    draw_ball b = do
      pattern_ball
      arc (ball_x b) (ball_y b) ball_radius 0 (2*pi)
      fill
    draw_penalty y = do
      setLineWidth 5.0
      pattern_penalty
      arc (win_width/2) y penalty_radius 0 (2*pi)
      strokePreserve
      
      setSourceRGB 0 0 0
      fill
      
      pattern_penalty
      arc (win_width/2) y (penalty_radius - 15) 0 (2*pi)
      strokePreserve
      
      setSourceRGB 0 0 0
      fill
    draw_score b c r = do
      show_text_centering ("Blue: " ++ (show b) ++ 
                           " Ball: " ++ (show c) ++ 
                           " Red: " ++ (show r)) 20
    show_text_centering t y = do
      te <- textExtents t
      moveTo ((win_width - textExtentsWidth te)/2) y
      showText t

cb_key_press_event window gameinfo = do 
  keyName <- eventKeyName
  [] <- eventModifier
  liftIO $ case keyName of 
    "r" -> modifyIORef gameinfo (restart)
    "a" -> modifyIORef gameinfo (update_gi Blue (-1))
    "s" -> modifyIORef gameinfo (update_gi Blue 1)
    "Left" ->  modifyIORef gameinfo (update_gi Red (-1))
    "Right" -> modifyIORef gameinfo (update_gi Red 1)
    "Escape" -> mainQuit
    _   -> return ()
  where
    restart gi@GameInfo{gi_state=state} 
      | (state ==  GS_WON_BY_RED || state == GS_WON_BY_BLUE) = gameinfo_new
      | otherwise = gi
    update_gi Blue vx gi@GameInfo{gi_slime_blue=sb@Slime{}} = 
      gi{gi_slime_blue=sb{slime_vx=vx}}
    update_gi Red vx gi@GameInfo{gi_slime_red=sr@Slime{}} = 
      gi{gi_slime_red=sr{slime_vx=vx}}

cb_timeout window gameinfo = do
  gi <- readIORef gameinfo
  case (gi_state gi) of
    GS_INIT -> modifyIORef gameinfo (update_state GS_SERVICE_BY_BLUE)
    GS_PLAY -> do
      modifyIORef gameinfo (update_slime Blue)
      modifyIORef gameinfo (update_slime Red)
      modifyIORef gameinfo (collision_ball_slime Blue)
      modifyIORef gameinfo (collision_ball_slime Red)
      modifyIORef gameinfo collision_penalty
      modifyIORef gameinfo update_ball
      modifyIORef gameinfo is_drop_ball
    GS_GOT_BY_BLUE -> modifyIORef gameinfo $ score_move Blue
    GS_GOT_BY_RED ->  modifyIORef gameinfo $ score_move Red
    GS_SERVICE_BY_BLUE -> modifyIORef gameinfo $ serve_set Blue
    GS_SERVICE_BY_RED  -> modifyIORef gameinfo $ serve_set Red
    GS_WON_BY_BLUE -> return ()
    GS_WON_BY_RED -> return ()
  widgetQueueDraw window
  return True
  where
    collision_ball_slime t gi@GameInfo{gi_ball=b@Ball{},
                                      gi_ball_count=count,
                                      gi_slime_blue=sb,
                                      gi_slime_red=sr} = 
      let s = if t == Blue then sb else sr
          v_s = (ball_x b, ball_y b)
          v_a = ((slime_x s) - (ball_x b), win_height - (ball_y b))
          v_b = ((slime_x s) - ((ball_x b) + (ball_vx b)), 
                 win_height - ((ball_y b)+(ball_vy b)))
          c = (inner_product v_a v_s) * (inner_product v_b v_s)
          d = (abs (outer_product v_s v_a) ) / norm(v_s)
          r = slime_radius + ball_radius in
      if d <= r && (c <= 0 || r > norm(v_a) || r > norm(v_b)) then
        let dx_0 = (ball_x b) - (slime_x s)
            dy_0 = (ball_y b) - win_height
            v = (dx_0, dy_0)
            dist_0 = norm(v)
            angle = acos(dx_0/dist_0)
            cos2x = cos(2*angle)
            sin2x = sin(2*angle) in
        gi{gi_ball=b{ball_vx =  (- ball_vx b) * cos2x + (ball_vy b) * sin2x,
                     ball_vy =   (ball_vx b) * sin2x + (ball_vy b) * cos2x},
           gi_ball_owner = t,
           gi_ball_count = count + 1}
      else
        gi
    collision_penalty gi@GameInfo{gi_ball=b@Ball{},
                                  gi_penalty_y = penalty_y,
                                  gi_score_blue = score_blue,
                                  gi_score_red = score_red,
                                  gi_ball_owner = owner,
                                  gi_wait_count = wait_count} = 
      let v = ((ball_x b) - win_width / 2, (ball_y b) - penalty_y) in
      if norm(v) < penalty_radius + ball_radius then
        reduce_score owner gi{ gi_wait_count= (wait_count -1)}
      else
        gi
      where
        reduce_score Blue gi@GameInfo{gi_score_blue=score,
                                      gi_wait_count=wait_count} = 
          gi{gi_score_blue=(score-1), gi_wait_count=4}
        reduce_score Red gi@GameInfo{gi_score_red=score,
                                      gi_wait_count=wait_count} = 
          gi{gi_score_red=(score-1), gi_wait_count=4}

    norm (a,b) = sqrt(a*a + b*b)
    inner_product (a,b) (c,d) = a * c + b * d
    outer_product (a,b) (c,d) = a * d - c * b
    update_state state gi@GameInfo{} = gi{gi_state=state}
    update_slime Blue gi@GameInfo{gi_slime_blue=sb@Slime{}} = 
      gi{gi_slime_blue=(slime_move sb)}
    update_slime Red gi@GameInfo{gi_slime_red=sr@Slime{}} = 
      gi{gi_slime_red=(slime_move sr)}

    update_ball gi@GameInfo{gi_ball=b@Ball{}} = 
      gi{gi_ball=(ball_move b)}
    is_drop_ball gi@GameInfo{gi_ball=b@Ball{ball_x=x, ball_y=y}} = 
      if y >= win_height - ball_radius then
        if x < win_width / 2 then
          gi{gi_state = GS_GOT_BY_RED}
        else
          gi{gi_state = GS_GOT_BY_BLUE}
      else
        gi
    serve_set team gi@GameInfo{gi_ball=ball,gi_slime_blue=sb,gi_slime_red=sr} = 
      update_state GS_PLAY gi{
        gi_ball = if team == Blue then
                    ball_new Blue
                  else
                    ball_new Red,
        gi_slime_blue=(slime_init sb),
        gi_slime_red=(slime_init sr),
        gi_penalty_y = 120, -- XXX
        gi_ball_count = 1,
        gi_wait_count = 0}
    score_move win gi@GameInfo{gi_ball_count=ball_count,
                           gi_wait_count=wait_count,
                           gi_score_blue=score_blue,
                           gi_score_red=score_red,
                           gi_state=state}
      | ball_count > 0 && wait_count > 0 = gi{gi_wait_count = (wait_count - 1)}
      | ball_count <= 0 && wait_count > 0 = gi{gi_wait_count = (wait_count - 1)}
      | ball_count <= 0 && wait_count <= 0 = 
        gi{gi_state = if win == Blue then 
                        GS_SERVICE_BY_BLUE 
                      else 
                        GS_SERVICE_BY_RED }
      | ball_count > 0 && wait_count <= 0 = 
        gi{gi_ball_count = (ball_count - 1),
           gi_wait_count = if ball_count == 0 then 50 else 7,
           gi_score_red = if win == Blue then score_red - 1 else score_red,
           gi_score_blue = if win == Red then score_blue - 1 else score_blue,
           gi_state = if win == Blue then
                         if score_red <= 0 then
                           GS_WON_BY_BLUE
                         else
                           state
                      else
                        if score_blue <= 0 then
                          GS_WON_BY_RED
                        else
                          state}
      
main = do
  initGUI
  gameinfo <- newIORef gameinfo_new
  window <- windowNew
  widgetSetSizeRequest window (round win_width) (round win_height)
  set window [ windowResizable := False]
  on window exposeEvent $ cb_expose_event gameinfo
  on window keyPressEvent $ tryEvent $ cb_key_press_event window gameinfo
  on window deleteEvent (liftIO $ (mainQuit >> return True))
  widgetShowAll window  
  timeoutAdd (cb_timeout window gameinfo) 10
  mainGUI

